---
title: "limma-voom DE analysis"
author: "Chi-Lam Poon"
output: html_document
---

```{R setup, include=T}
suppressPackageStartupMessages({
  library(purrr)
  library(dplyr)
  library(metap)
  library(knitr)
  library(ggplot2)
  library(ggpubr)
  library(Biobase)
})
```


#### Load the Rdata filted before
```{R}
main_dir <- "~/cvd/v6"
eset <- load(file.path(main_dir, "data/filted.cevd.rdata"))
```


### Differential Expression Analysis

Differential expression (DE) analysis between CeVD and non-CeVD samples was conducted using the voom-limma pipeline. We first did DE analyses on three brain clusters, then on other tissues.

For **brain clusters**, gene expression was modelled using a linear regression formula with blocking, which is due to multiple and various tissue sites in a cluster:

$$Y \sim \beta_1Sex + \beta_2Age + \beta_3BMI + \beta_4Hardy + \beta_5Batch + \beta_6MHCVD + \beta_7PC1_g + \beta_8PC2_g + \beta_9PC3_g + \beta_{10}Site + \mu_1Subject + \epsilon$$
* Note that no need to add race as a variable as it's similar to PC 1~3.

```{R}
table_dir <- file.path(main_dir, "results/limma.tables")
plot_dir <- file.path(main_dir, "results/limma.plots")
exprs <- exprs(filtedSet)
pdata <- pData(filtedSet)

clusters <- c("brain-0", "brain-1", "brain-2")
tissues <- unique(as.character(pdata$our_subtypes))
tissues <- tissues[!tissues %in% grep("brain", tissues, value = T)]
disease <- "cvd1"
```


```{R}
source("~/cvd/v6/scripts/utils.R")

for (c in clusters) {
  print(c)
  # Subset phenotype and expression data
  sub_pdata <- pdata[pdata$our_subtypes==c, ]
  sub_exprs <- exprs[, which(colnames(exprs) %in% sub_pdata$SAMPID)]
  
  # Construct the linear model
  print("Running limma pipeline...")
  age    <- as.numeric(sub_pdata$AGE)
  bmi    <- as.numeric(sub_pdata$BMI)
  gender <- factor(sub_pdata$gender)
  batch  <- factor(sub_pdata$SMNABTCHT)
  cvd    <- factor(sub_pdata$MHCVD)
  #race   <- factor(as.character(sub_pdata$RACE))
  subject<- factor(sub_pdata$SUBJID)
  hardy  <- factor(as.character(sub_pdata$DTHHRDY))
  site   <- factor(sub_pdata$SMTSD)
  PC1    <- as.numeric(sub_pdata$C1)
  PC2    <- as.numeric(sub_pdata$C2)
  PC3    <- as.numeric(sub_pdata$C3)
  design <- model.matrix(~  age + bmi + gender + batch + cvd + site + hardy + PC1 + PC2 + PC3)  
  
  # Use voom weighting
  transformedCounts <- sub_exprs
  voomOutput <- voomWeightsCustomized(transformedCounts, design)
  
  # Blocking
  corfit <- duplicateCorrelation(voomOutput, design, block = subject)
  print('Consensus correlation:')
  print(corfit$consensus.correlation)
  fit <- lmFit(voomOutput, design, block = subject, correlation = corfit$consensus.correlation)
  fit <- eBayes(fit)
  gl <- fData(filtedSet)$geneNames
  tt <- topTable(fit, number=Inf, genelist=gl, coef=disease)
  
  # Output reports & plots
  write.table(tt, file=sprintf("%s/%s.tsv", table_dir, c), sep="\t", quote=F)
  pdf(sprintf("%s/%s.volcanoplot.pdf", plot_dir, c))
  limma::volcanoplot(fit, coef=disease, highlight = 6, names = gl, main = c)
  dev.off()
  pdf(sprintf("%s/%s.MAplot.pdf", plot_dir, c))
  limma::plotMA(fit, main = c)
  dev.off()
}
```

For **other tissues**, gene expression was modelled using the following linear regression formula:

$$Y \sim \beta_1Sex + \beta_2Age + \beta_3BMI + \beta_4Hardy + \beta_5Batch + \beta_6MHCVD + \beta_7PC1_g + \beta_8PC2_g + \beta_9PC3_g + \epsilon$$
Only tissues with more than 10 or 15 CeVD samples should be used in DE analyses, but here just using all.

```{R, eval=T}
tissues <- unique(as.character(pdata$our_subtypes))
tissues <- tissues[!tissues %in% grep("brain", tissues, value = T)]

for (tissue in tissues) {
  print(tissue)
  
  # Subset phenotype and expression data
  sub_pdata <- pdata[pdata$our_subtypes==tissue, ]
  sub_exprs <- exprs[ ,which(colnames(exprs) %in% sub_pdata$SAMPID)]
  
  # Construct the linear model
  print("Running limma pipeline...")
  age    <- as.numeric(sub_pdata$AGE)
  bmi    <- as.numeric(sub_pdata$BMI)
  gender <- factor(sub_pdata$gender)
  batch  <- factor(sub_pdata$SMNABTCHT)
  cvd    <- factor(sub_pdata$MHCVD)
  #race   <- factor(as.character(sub_pdata$RACE))
  hardy  <- factor(as.character(sub_pdata$DTHHRDY))
  PC1    <- as.numeric(sub_pdata$C1)
  PC2    <- as.numeric(sub_pdata$C2)
  PC3    <- as.numeric(sub_pdata$C3)
  design <- model.matrix(~  age + bmi + gender + batch + cvd + hardy + PC1 + PC2 + PC3)  
  
  # Use voom weighting
  transformedCounts <- sub_exprs
  voomOutput <- voomWeightsCustomized(transformedCounts, design)
  fit <- lmFit(voomOutput, design)
  fit <- eBayes(fit)
  gl <- fData(filtedSet)$geneNames
  tt <- topTable(fit, number=Inf, genelist=gl, coef=disease)
  
  # Output reports & plots
  write.table(tt, file=sprintf("%s/%s.tsv", table_dir, tissue), sep="\t", quote=F)
  pdf(sprintf("%s/%s.volcanoplot.pdf", plot_dir, tissue))
    limma::volcanoplot(fit, coef=disease, highlight = 6, names = gl, main = tissue)
  dev.off()
  pdf(sprintf("%s/%s.MAplot.pdf", plot_dir, tissue))
    limma::plotMA(fit, main = tissue)
  dev.off()
}
```


#### Rank the genes in three brain clusters
For brain clusters, genes from limma reports were ranked by calculating the combined Q-value from the adjusted P-values in three clusters using Fisherâ€™s method, finding the common set of genes which share similar expression patterns across brain regions.

```{R}
btt0 <- read.table(file.path(table_dir, "brain-0.tsv"), sep="\t", header=T)
btt1 <- read.table(file.path(table_dir, "brain-1.tsv"), sep="\t", header=T)
btt2 <- read.table(file.path(table_dir, "brain-2.tsv"), sep="\t", header=T)
btt0$esID <- rownames(btt0)
btt1$esID <- rownames(btt1)
btt2$esID <- rownames(btt2)

# Merge three adjusted p-vals
cb_pval <- list(btt0[ ,c("esID", "ID", "t", "adj.P.Val")], btt1[ ,c("esID", "t", "adj.P.Val")], btt2[ ,c("esID", "t", "adj.P.Val")]) %>% reduce(left_join, by="esID")

# Compute combined q-val and sort
cb_pval$fisherQ <- sapply(seq_along(cb_pval$esID), function(x) sumlog(c(cb_pval$adj.P.Val.x[x], cb_pval$adj.P.Val.y[x], cb_pval$adj.P.Val[x]))$p)
cb_pval <- cb_pval[order(cb_pval$fisherQ), ]
head(cb_pval, 10)

# Save 
write.table(cb_pval, file.path(main_dir, "results/combined.brain.p.txt"), sep="\t", col.names = T, row.names = F, quote = F)
```


### Session Information
```{R}
sessionInfo()
```