---
title: "limma-voom"
author: "Chi-Lam Poon"
output: html_document
---

```{R set up, include=T}
suppressPackageStartupMessages({
  library(purrr)
  library(dplyr)
  library(metap)
  library(knitr)
  library(ggplot2)
  library(ggpubr)
  library(Biobase)
})
```

#### Load the Rdata filted before
```{R}
main_dir <- "/home/ubuntu/hdd/cevd"
eset <- load(file.path(main_dir, "rdata/filted.cevd.rdata"))
```

### Demography

##### MHCVD 0 & 1 in each tissue
There are total 6295 samples in the filted dataset, among these 523 had cerebrovascular disease histories (MHCVD = 1) while 5772 didn't (MHCVD = 0).

```{R, fig.height=10, fig.width=10}
pdata <- pData(filtedSet)
pdata$MHCVD <- as.factor(pdata$MHCVD)
m_count1 <- pdata %>% group_by(MHCVD) %>% summarise(counts = n())
m_count1$MHCVD <- as.factor(m_count1$MHCVD)
m_count2 <- pdata %>% group_by(our_subtypes, MHCVD) %>% summarise(counts = n())
m_count2$MHCVD <- as.factor(m_count2$MHCVD)
m_count2$our_subtypes <- as.factor(gsub("_", " ", m_count2$our_subtypes))

m1 <- ggplot(m_count1, aes(x=MHCVD, y=counts, fill=MHCVD)) + 
  geom_bar(stat = "identity") + 
  geom_text(aes(label=counts), vjust = -0.1, size = 3.2) +
  theme_minimal() + theme(axis.title.x=element_blank()) + 
  scale_fill_manual(values = c("#1aa398", "#ffb353"))

m2 <- ggplot(m_count2, aes(x=MHCVD, y=counts, fill=MHCVD)) +
  geom_bar(stat="identity") + 
  geom_text(aes(label=counts), vjust = 0.5, size = 3.2) + 
  theme_minimal() +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.title.y=element_blank()) +
  scale_fill_manual(values = c("#1aa398", "#ffb353")) +
  facet_wrap(~our_subtypes,ncol=4)

ggarrange(m1, m2, ncol=2, nrow=1, widths=c(1,3), common.legend = T, legend = "right")
```


##### Subject information comparison in gender, AGE, RACE and BMI

```{R}
# Total subject number
pdata2 <- pdata %>% select(SUBJID, gender, AGE, RACE, BMI, MHCVD)
rownames(pdata2) <- seq(1, dim(pdata2)[1])
pdata2 <- unique(pdata2)
dim(pdata2)[1]
table(pdata2$MHCVD)
```
Total 444 subjects, and 41 subjects with CeVD histories while 387 without.

##### Age

```{R}
# Function for T-test
calculateT <- function(v1, v2){
  v1 <- as.numeric(v1)
  v2 <- as.numeric(v2)
  Ftest <- var.test(v1, v2, alternative = "two.sided")
  Ttest <- NA
  ifelse(Ftest$p.value <= 0.05, Ttest <- t.test(v1, v2, paired = F, var.equal = F),
    Ttest <- t.test(v1, v2, paired = F, var.equal = T))
  
  Ttest
}

# Mean and standard deviation for ages of non-CeVD subjects
mean(pdata2[pdata2$MHCVD==0, ]$AGE)
sd(pdata2[pdata2$MHCVD==0, ]$AGE)
# Mean and standard deviation for ages of CeVD subjects
mean(pdata2[pdata2$MHCVD==1, ]$AGE)
sd(pdata2[pdata2$MHCVD==1, ]$AGE)
# T-test for ages in two cohorts
t.age <- calculateT(pdata2[pdata2$MHCVD==0, ]$AGE, pdata2[pdata2$MHCVD==1, ]$AGE)
t.age
age_sum <- pdata2 %>%
  select(AGE, MHCVD) %>%
  group_by(MHCVD) %>%
  summarise(mean.age = mean(AGE, na.rm = T))
m_a1 <- ggplot(pdata2, aes(AGE, fill=MHCVD)) +
  geom_histogram(aes(y=..count..), alpha=0.5) +
  geom_vline(data=age_sum, aes(xintercept=mean.age, colour=MHCVD), lty=2, size=1) +
  scale_fill_brewer(palette="Set2") +
  scale_colour_brewer(palette="Set2") +
  scale_y_continuous(labels=scales::comma) +
  ylab("Density") +
  ggtitle(paste0("Age distribution in two cohorts    P-val = ", formatC(t.age$p.value, format="e", digits = 2))) + 
  labs(x="ages") +
  theme_minimal() 
m_a2 <- ggplot(pdata2, aes(AGE, fill=MHCVD)) +
  geom_histogram(aes(y=..density..), alpha=0.5) +
  geom_density(alpha=.2, aes(colour=MHCVD)) +
  geom_vline(data=age_sum, aes(xintercept=mean.age, colour=MHCVD), lty=2, size=1) +
  scale_fill_brewer(palette="Set2") +
  scale_colour_brewer(palette="Set2") +
  scale_y_continuous(labels=scales::percent) +
  ylab("Density") +
  ggtitle("") + 
  labs(x="ages") +
  theme_minimal() 
ggarrange(m_a1, m_a2, ncol=2, nrow=1, common.legend = T, legend = "right")
```

##### BMI

```{R}
# Mean and standard deviation for BMIs of non-CeVD subjects
mean(pdata2[pdata2$MHCVD==0, ]$BMI)
sd(pdata2[pdata2$MHCVD==0, ]$BMI)
# Mean and standard deviation for BMIs of CeVD subjects
mean(pdata2[pdata2$MHCVD==1, ]$BMI)
sd(pdata2[pdata2$MHCVD==1, ]$BMI)
# T-test for BMI in two cohorts
t.bmi <- calculateT(pdata2[pdata2$MHCVD==0, ]$BMI, pdata2[pdata2$MHCVD==1, ]$BMI)
t.bmi
bmi_sum <- pdata2 %>%
  select(BMI, MHCVD) %>%
  group_by(MHCVD) %>%
  summarise(mean.bmi = mean(BMI, na.rm = T))
b_a1 <- ggplot(pdata2, aes(BMI, fill=MHCVD)) +
  geom_histogram(aes(y=..count..), alpha=0.5) +
  geom_vline(data=bmi_sum, aes(xintercept=mean.bmi, colour=MHCVD), lty=2, size=1) +
  scale_fill_brewer(palette="Set3") +
  scale_colour_brewer(palette="Set3") +
  scale_y_continuous(labels=scales::comma) +
  ylab("Density") +
  ggtitle(paste0("BMI distribution in two cohorts    P-val = ", round(t.bmi$p.value, 2))) + 
  labs(x="BMI") +
  theme_minimal() 
b_a2 <- ggplot(pdata2, aes(BMI, fill=MHCVD)) +
  geom_histogram(aes(y=..density..), alpha=0.5) +
  geom_density(alpha=.2, aes(colour=MHCVD)) +
  geom_vline(data=bmi_sum, aes(xintercept=mean.bmi, colour=MHCVD), lty=2, size=1) +
  scale_fill_brewer(palette="Set3") +
  scale_colour_brewer(palette="Set3") +
  scale_y_continuous(labels=scales::percent) +
  ylab("Density") +
  ggtitle("") + 
  labs(x="BMI") +
  theme_minimal() 
ggarrange(b_a1, b_a2, ncol=2, nrow=1, common.legend = T, legend = "right")
```

##### Gender

```{R}
table(pdata2[pdata2$MHCVD==1, ]$gender)
table(pdata2[pdata2$MHCVD==0, ]$gender)
chi.sex <- chisq.test(rbind(table(pdata2[pdata2$MHCVD==1, ]$gender), table(pdata2[pdata2$MHCVD==0, ]$gender)))
chi.sex
m_g1 <- ggplot(pdata2, aes(gender, fill=MHCVD)) +
  geom_bar(position="stack") +
  scale_fill_brewer(palette="Set1") + 
  scale_y_continuous(labels = scales::comma) +
  ylab("MHCVD") + 
  ggtitle(paste0("MHCVD by Sex    P-val = ", round(chi.sex$p.value, 2))) +
  theme_minimal()

m_g2 <- ggplot(pdata2, aes(gender, fill=MHCVD)) +
  geom_bar(position="fill") +
  scale_fill_brewer(palette="Set1") + 
  scale_y_continuous(labels = scales::percent) +
  ylab("MHCVD Rate") + 
  ggtitle("") +
  theme_minimal()

ggarrange(m_g1, m_g2, ncol=2, nrow=1, common.legend = T, legend = "right")
```

##### Race

```{R}
pdata2$RACE <- as.factor(pdata2$RACE)
table(pdata2[pdata2$MHCVD==1, ]$RACE)
table(pdata2[pdata2$MHCVD==0, ]$RACE)
chi.race <- chisq.test(rbind(table(pdata2[pdata2$MHCVD==1, ]$RACE), table(pdata2[pdata2$MHCVD==0, ]$RACE)))
chi.race

m_r1 <- ggplot(pdata2, aes(RACE, fill=MHCVD)) +
  geom_bar(position="stack") +
  scale_fill_brewer(palette="Accent") + 
  scale_x_discrete(labels = c("Asian", "African", "American", "White", "Unknown")) +
  scale_y_continuous(labels = scales::comma) +
  ylab("MHCVD") + 
  ggtitle(paste0("MHCVD by Race    P-val = ", round(chi.race$p.value, 2))) +
  theme_minimal() + theme(axis.text.x = element_text(angle = 55, hjust = 1, size = 8.5))

m_r2 <- ggplot(pdata2, aes(RACE, fill=MHCVD)) +
  geom_bar(position="fill") +
  scale_fill_brewer(palette="Accent") + 
    scale_x_discrete(labels = c("Asian", "African", "American", "White", "Unknown")) +
  scale_y_continuous(labels = scales::percent) +
  ylab("MHCVD Rate") + 
  ggtitle("") +
  theme_minimal() + theme(axis.text.x = element_text(angle = 55, hjust = 1, size = 8.5))

ggarrange(m_r1, m_r2, ncol=2, nrow=1, common.legend = T, legend = "right")
```

##### Tissues in Brian-0,1,2

```{R}
for(i in c("brain-0", "brain-1", "brain-2")){
  brain_sum <- pdata %>% filter(our_subtypes==i) %>%
    select(SAMPID, SMTSD, MHCVD) %>%
    group_by(SMTSD, MHCVD) %>%
    summarise(n = n()) 
  print(kable(brain_sum, caption="tissues in cluster", format="markdown"))
}
```

### Differential Expression Analysis

Differential expression (DE) analysis between CeVD and non-CeVD samples was conducted using the voom-limma pipeline. We first conducted DE analyses on three brain clusters, then on other tissues.

For **brain clusters**, gene expression was modelled using a linear regression formula with blocking, which is due to multiple and various tissue sites in a cluster:

$$Y \sim \beta_1Sex + \beta_2Age + \beta_3BMI + \beta_4Hardy + \beta_5Batch + \beta_6MHCVD + \beta_7PC1_g + \beta_8PC2_g + \beta_9PC3_g + \beta_{10}Site + \mu_1Subject + \epsilon$$

```{R}
table_dir <- file.path(main_dir, "results/limma.tables")
plot_dir <- file.path(main_dir, "results/limma.plots")
exprs <- exprs(filtedSet)
clusters <- c("brain-0", "brain-1", "brain-2")
tissues <- unique(as.character(pdata$our_subtypes))
tissues <- tissues[!tissues %in% grep("brain", tissues, value = T)]
disease <- "cvd1"
```


```{R, eval=T}
source("/home/rstudio/cevd/utils.R")
for(c in clusters){
  print(c)
  
  # Subset phenotype and expression data
  sub_pdata <- pdata[pdata$our_subtypes==c, ]
  sub_exprs <- exprs[, which(colnames(exprs) %in% sub_pdata$SAMPID)]
  # Construct the linear model
  print("Running limma pipeline...")
  age    <- sub_pdata$AGE
  bmi    <- sub_pdata$BMI
  gender <- sub_pdata$gender
  batch  <- sub_pdata$SMNABTCHT
  cvd    <- factor(sub_pdata$MHCVD)
  #race   <- factor(as.character(sub_pdata$RACE))
  subject<- sub_pdata$SUBJID
  hardy  <- factor(as.character(sub_pdata$DTHHRDY))
  site   <- as.character(sub_pdata$SMTSD)
  PC1    <- sub_pdata$C1
  PC2    <- sub_pdata$C2
  PC3    <- sub_pdata$C3
  design = model.matrix(~  age + bmi + gender + batch + cvd + site + hardy + PC1 + PC2 + PC3)  
  
  # Use voom weighting
  transformedCounts <- sub_exprs
  voomOutput <- voomWeightsCustomized(transformedCounts, design)
  
  # Blocking
  corfit <- duplicateCorrelation(voomOutput, design, block = subject)
  print('Consensus correlation:')
  print(corfit$consensus.correlation)
  fit <- lmFit(voomOutput, design, block = subject, correlation = corfit$consensus.correlation)
  fit <- eBayes(fit)
  gl <- fData(filtedSet)$geneNames
  tt <- topTable(fit, number=Inf, genelist=gl, coef=disease)
  
  # Output reports & plots
  write.table(tt, file=sprintf("%s/%s.tsv", table_dir, c), sep="\t", quote=F)
  pdf(sprintf("%s/%s.volcanoplot.pdf", plot_dir, c))
  limma::volcanoplot(fit, coef=disease, highlight = 6, names = gl, main = c)
  dev.off()
  pdf(sprintf("%s/%s.MAplot.pdf", plot_dir, c))
  limma::plotMA(fit, main = c)
  dev.off()
}

```

For **other tissues**, gene expression was modelled using the following linear regression formula:

$$Y \sim \beta_1Sex + \beta_2Age + \beta_3BMI + \beta_4Hardy + \beta_5Batch + \beta_6MHCVD + \beta_7PC1_g + \beta_8PC2_g + \beta_9PC3_g + \epsilon$$
Only tissues with more than 10 or 15 CeVD samples should be used in DE analyses, but here just using all.

```{R, eval=T}
tissues <- unique(as.character(pdata$our_subtypes))
tissues <- tissues[!tissues %in% grep("brain", tissues, value = T)]
for(tissue in tissues){
  print(tissue)
  
  # Subset phenotype and expression data
  sub_pdata <- pdata[pdata$our_subtypes==tissue, ]
  sub_exprs <- exprs[, which(colnames(exprs) %in% sub_pdata$SAMPID)]
  # Construct the linear model
  print("Running limma pipeline...")
  age    <- sub_pdata$AGE
  bmi    <- sub_pdata$BMI
  gender <- sub_pdata$gender
  batch  <- sub_pdata$SMNABTCHT
  cvd    <- factor(sub_pdata$MHCVD)
  #race   <- factor(as.character(sub_pdata$RACE))
  hardy  <- factor(as.character(sub_pdata$DTHHRDY))
  PC1    <- sub_pdata$C1
  PC2    <- sub_pdata$C2
  PC3    <- sub_pdata$C3
  design = model.matrix(~  age + bmi + gender + batch + cvd + hardy + PC1 + PC2 + PC3)  
  
  # Use voom weighting
  transformedCounts <- sub_exprs
  voomOutput <- voomWeightsCustomized(transformedCounts, design)
  fit <- lmFit(voomOutput, design)
  fit <- eBayes(fit)
  gl <- fData(filtedSet)$geneNames
  tt <- topTable(fit, number=Inf, genelist=gl, coef=disease)
  
  # Output reports & plots
  write.table(tt, file=sprintf("%s/%s.tsv", table_dir, tissue), sep="\t", quote=F)
  pdf(sprintf("%s/%s.volcanoplot.pdf", plot_dir, tissue))
    limma::volcanoplot(fit, coef=disease, highlight = 6, names = gl, main = tissue)
  dev.off()
  pdf(sprintf("%s/%s.MAplot.pdf", plot_dir, tissue))
    limma::plotMA(fit, main = tissue)
  dev.off()
}
```


#### Rank the genes in three brain clusters

For brain clusters, genes from limma reports were ranked by calculating the combined Q-value from the adjusted P-values in three clusters using Fisher’s method, finding the common set of genes which share similar expression patterns across brain regions.

```{R}
btt0 <- read.table(file.path(table_dir, "brain-0.tsv"), sep="\t", header=T)
btt1 <- read.table(file.path(table_dir, "brain-1.tsv"), sep="\t", header=T)
btt2 <- read.table(file.path(table_dir, "brain-2.tsv"), sep="\t", header=T)
btt0$esID <- rownames(btt0)
btt1$esID <- rownames(btt1)
btt2$esID <- rownames(btt2)

# Merge three adjusted p-vals
cb_df <- list(btt0[, c("esID", "ID", "t", "adj.P.Val")], btt1[, c("esID", "t", "adj.P.Val")], btt2[, c("esID", "t", "adj.P.Val")]) %>% reduce(left_join, by="esID")

# Compute combined q-val and sort
cb_df$fisherQ <- sapply(seq_along(cb_df$esID), function(x) sumlog(c(cb_df$adj.P.Val.x[x], cb_df$adj.P.Val.y[x], cb_df$adj.P.Val[x]))$p)
cb_df <- cb_df[order(cb_df$fisherQ), ]
head(cb_df, 10)

# Save 
write.table(cb_df, file.path(main_dir, "results/combined.brains.txt"), sep="\t", col.names = T, row.names = T, quote = F)
```

### Session Information

```{R}
sessionInfo()
```
