---
title: "Plots for DEGs"
author: "Chi-Lam Poon"
output: html_document
---

```{R setup, include = T}
suppressPackageStartupMessages({
  library(purrr)
  library(dplyr)
  library(metap)
  library(knitr)
  library(ggplot2)
  library(ggpubr)
  library(Biobase)
  library(RColorBrewer)
  library(pheatmap)
  library(venn)
})
```

### Load files

```{R}
main_dir <- "/home/ubuntu/hdd/cevd"
table_dir <- file.path(main_dir, "results/limma.tables")
plot_dir <- file.path(main_dir, "results/limma.plots")

eset <- load(file.path(main_dir, "rdata/filted.cevd.rdata"))
exprs <- exprs(filtedSet)
pdata <- pData(filtedSet)
anno <- fData(filtedSet)
```

#### DEG amounts in each tissue or cluster
```{R, fig.height=8}
clusters <- c("brain-0", "brain-1", "brain-2")
tissues <- unique(as.character(pdata$our_subtypes))
tissues <- tissues[!tissues %in% grep("brain", tissues, value = T)]
all_types <- c(clusters, tissues)


# A function to plot DEG numbers for all tissues
count_DEG <- function(table_dir, all_types){
  sig_count <- c()
  for(i in all_types){
    tempt <- read.table(file.path(table_dir, paste0(i, ".tsv")), sep="\t", header=T)
    sig_count <- c(sig_count, nrow(tempt[tempt$adj.P.Val < 0.1 & abs(tempt$logFC) > log2(1.5), ]))
  }
  
  sc_df <- data.frame(type = all_types, count = sig_count)
  
  ggplot(data=sc_df, aes(x=type, y=count)) +
    geom_bar(stat="identity", fill="skyblue", color="black") +
    scale_x_discrete(limits=rev(sc_df$type), labels=rev(gsub("_", " ", sc_df$type))) +
    scale_y_continuous(limits=c(0,1280), expand = c(0.02,0)) +
    geom_text(aes(label=count), vjust=0.25, hjust=-0.3, size=2.7) +
    labs(title="DEGs amounts", x="", y="") +
    coord_flip() + theme_bw()
}
count_DEG(table_dir, all_types)
```





#### Heatmaps

```{R heatmap, fig.height=13, fig.width=16}

# Create a color vector containing at most 74 colour hexcodes
get_colVector <- function(n){
  col <- brewer.pal.info[brewer.pal.info$category=='qual',] # get max. 74 colours
  col_vector = unlist(mapply(brewer.pal, col$maxcolors, rownames(col)))
  ifelse(n > length(col_vector),
        vec <- sample(col_vector, n, replace=T), 
        vec <- sample(col_vector, n, replace=F))
  vec
}

# Function to draw heatmap for top 100 DEGs with q-val < 0.1 & logFC > 1.5
draw_heatmap <- function(tissue, clusters, exprs, pdata){
  # Load toptable
  tt <- read.delim(file.path(table_dir, paste0(tissue, ".tsv")), sep="\t", header=T)
  tt <- tt[tt$adj.P.Val < 0.1 & abs(tt$logFC) > log2(1.5), ]
  if(dim(tt)[1] > 100){
    tt <- tt[1:100, ]
  }
  hm_pdata <- pdata[pdata$our_subtypes==tissue, ] 
  hm_pdata$MHCVD <- factor(hm_pdata$MHCVD, levels = c(0, 1)) # reorder MHCVD
  hm_exprs <- exprs[which(rownames(exprs) %in% rownames(tt)), which(colnames(exprs) %in% rownames(hm_pdata))]
  hm_gene <- tt[match(rownames(hm_exprs), rownames(tt)),]$ID
  
  # Assign the annotation colours
  MHCVD <- c("#464646", "#d3d3d2")
  names(MHCVD) <- c(0, 1) # only for 0 & 1
  
  if(tissue %in% clusters){
    hm_pdata <- hm_pdata %>% select(SMTSD, MHCVD)
    hm_pdata$SMTSD <- gsub('^.* - ', '', hm_pdata$SMTSD)
    smtsd <- unique(as.character(hm_pdata$SMTSD))
    SMTSD <- get_colVector(length(smtsd))
    names(SMTSD) <- smtsd
    anno_colors <- list(SMTSD  = SMTSD,
                        MHCVD = MHCVD) # the name must be consistent
  }else{
    hm_pdata <- hm_pdata %>% select(MHCVD)
    anno_colors <- list(MHCVD = MHCVD)
  }
  
  # Draw heatmap
  h <- pheatmap(hm_exprs, annotation_col=hm_pdata, annotation_colors=anno_colors, 
                labels_row = hm_gene, show_colnames = F)
  h
}


draw_heatmap("brain-0", clusters, exprs, pdata)
draw_heatmap("brain-1", clusters, exprs, pdata)
draw_heatmap("brain-2", clusters, exprs, pdata)
draw_heatmap("heart_atrial_appendage", clusters, exprs, pdata)
#draw_heatmap("minor_salivary_gland", clusters, exprs, pdata)
```

##### Heatmaps for three brain clusters
###### Use DEGs with combined Fisher's q-val < 0.05

```{R, fig.height=13, fig.width=16}
cb_df <- read.delim(file.path(main_dir, "results/combined.brains.txt"), sep="\t", header=T, row.names = 1)

big_heatmap <- function(cb_df, clusters, exprs, pdata){
  # Select those with q-val < 0.05
  cbDF <- cb_df[cb_df$fisherQ < 0.05, ]
  hm_pdata <- pdata[pdata$our_subtypes %in% clusters, ] %>% select(our_subtypes, MHCVD)
  hm_pdata$MHCVD <- factor(hm_pdata$MHCVD, levels = c(0, 1)) # reorder MHCVD
  hm_pdata$our_subtypes <- factor(hm_pdata$our_subtypes)
  hm_exprs <- exprs[which(rownames(exprs) %in% cbDF$esID), which(colnames(exprs) %in% rownames(hm_pdata))]
  hm_gene <- cbDF[match(rownames(hm_exprs), cbDF$esID),]$ID
  
  # Assign the annotation colours
  MHCVD <- c("#464646", "#d3d3d2")
  names(MHCVD) <- c(0, 1) # only for 0 & 1
  our_subtypes <- get_colVector(length(clusters))
  names(our_subtypes) <- clusters
  
  anno_colors <- list(our_subtypes = our_subtypes,
                      MHCVD = MHCVD)

  
  # Draw heatmap
  h <- pheatmap(hm_exprs, annotation_col=hm_pdata, annotation_colors=anno_colors, labels_row = hm_gene, show_colnames = F)
  h
}

big_heatmap(cb_df, clusters, exprs, pdata)

#big_heatmap(cb_df, "brain-0", exprs, pdata)


```



### Box plots of gene expression

##### Brain clusters
Boxplots for brain transcriptomes, the left one is the overall expression level in all brain tissues (including 3 clusters), the right facet plots are expression level in each single cluster.

```{R, fig.height=7, fig.width=9}
# Boxplot for gene expression level comparison in overall or tissue-specific way.
brain_box <- function(exprs, cb_df, pdata, esid, clusters){
  
  t_pdata <- pdata %>% select("SAMPID", "our_subtypes", "MHCVD") %>% filter(our_subtypes %in% clusters)
  exp_list <- as.data.frame(exprs[esid, which(colnames(exprs) %in% t_pdata$SAMPID)])
  exp_list$SAMPID <- rownames(exp_list)
  colnames(exp_list) <- c("counts", "SAMPID")
  mdf <- merge(exp_list, t_pdata, by="SAMPID")
  mdf$MHCVD <- factor(mdf$MHCVD)
  symbol <- as.character(cb_df[cb_df$esID==esid, ]$ID)
  q_val <- cb_df[cb_df$esID==esid, ]$fisherQ
  
  p1 <- ggplot(data=mdf, aes(x=MHCVD, y=counts)) + 
    geom_boxplot(aes(fill=MHCVD))  + 
    scale_fill_manual(values = c("steelblue2", "wheat1")) + # box fill color
    labs(title=symbol, x="", y="qsmooth counts") + guides(fill=F) +
    theme_bw()
  
  p2 <- ggboxplot(mdf, x="MHCVD", y="counts",
                 color="MHCVD", palette="jco", main=paste0("Combined q-val = ", formatC(q_val, format="e", digits=2)),
                 xlab="", ylab="", add="jitter") +
      theme_bw() +
      theme(strip.background = element_rect(colour = "black", fill = "white")) +
      facet_wrap(~our_subtypes)  + 
      stat_compare_means() # wilcoxon rank sum test
  
  ggarrange(p1, p2, ncol=2, nrow=1, widths=c(1,3))
}

# FCGBP
brain_box(exprs, cb_df, pdata, "ENSG00000090920.9", clusters)
# ANKRD37 
brain_box(exprs, cb_df, pdata, "ENSG00000186352.4", clusters)
# BEND3 
brain_box(exprs, cb_df, pdata, "ENSG00000178409.9", clusters)
# STAB1 
brain_box(exprs, cb_df, pdata, "ENSG00000010327.6", clusters)

```


#### DEGs in heart atrial appendage

Since there are over 1,000 DEGs found in heart atrial appendage, it's worth having a deeper look at these genes.
But how to prove that these are not due to the unbalanced cohort numbers...?

```{R, fig.width=4}
aa_tt <- read.table(file.path(table_dir, "heart_atrial_appendage.tsv"), sep="\t", header=T)
head(aa_tt, 20)

# Function to draw the boxplot for a single gene in one tissue
single_box <- function(countsMat, phenoDF, esid, tissue){
  # Load the top table of this tissue
  tt <- read.table(sprintf("%s/%s.tsv", table_dir, tissue), sep="\t", header=T)
  
  t_pdata <- phenoDF %>% filter(our_subtypes == tissue) %>% select("SAMPID", "MHCVD") 
  exp_list <- as.data.frame(countsMat[esid, which(colnames(countsMat) %in% t_pdata$SAMPID)])
  exp_list$SAMPID <- rownames(exp_list)
  colnames(exp_list) <- c("counts", "SAMPID")
  mdf <- merge(exp_list, t_pdata, by="SAMPID")
  mdf$MHCVD <- factor(mdf$MHCVD)
  symbol <- as.character(tt[esid, ]$ID)
  q_val <- tt[esid, ]$adj.P.Val
  
  ggboxplot(mdf, x="MHCVD", y="counts",
           color="MHCVD", palette="jco", main=paste0(symbol, "  q-val = ", formatC(q_val, format="e", digits=2)),
           xlab=tissue, ylab="qsmooth counts", add="jitter", ggtheme = theme_bw())
  
}

single_box(exprs, pdata, "ENSG00000151117.4", "heart_atrial_appendage")
single_box(exprs, pdata, "ENSG00000231769.2", "heart_atrial_appendage")
single_box(exprs, pdata, "ENSG00000004534.10", "heart_atrial_appendage")
```

### Overlap genes between RAA & brains
```{R overlap}
# Overlapp between genes from 3 clusters with < 0.05 Fisher's Q and those from RAA with q-val < 0.05
gene_brain <- as.vector(cb_df[cb_df$fisherQ < 0.05,]$ID)
gene_raa <- as.vector(aa_tt[aa_tt$adj.P.Val < 0.05,]$ID)

ven1 <- venn(list(Brains=gene_brain, RAA=gene_raa), intersections = T)
isect1 <- attr(ven1, 'intersection'); isect1$`Brains:RAA`

# # Overlapp between genes from each cluster with q-val < 0.1
btt0 <- read.table(file.path(table_dir, "brain-0.tsv"), sep="\t", header=T)
btt1 <- read.table(file.path(table_dir, "brain-1.tsv"), sep="\t", header=T)
btt2 <- read.table(file.path(table_dir, "brain-2.tsv"), sep="\t", header=T)

gene0 <- as.vector(btt0[btt0$adj.P.Val < 0.1 & abs(btt0$logFC) > log2(1.5),]$ID)
gene1 <- as.vector(btt1[btt1$adj.P.Val < 0.1 & abs(btt1$logFC) > log2(1.5),]$ID)
gene2 <- as.vector(btt2[btt2$adj.P.Val < 0.1 & abs(btt2$logFC) > log2(1.5),]$ID)

ven2 <- venn(list(Brain0=gene0, Brain1=gene1, Brain2=gene2), intersections = T)
```

