---
title: "ExpressionSet for GTEx v8"
author: "Chi-Lam Poon"
date: "12/15/2019"
output: html_document
---

Genotype-Tissue Expression (GTEx) project v8phs000424.v8.p2 was released on 2019-08-26. All files were downloaded from [GTEx portal](https://gtexportal.org/home/datasets) on 09/12/2019.

The encrypted files of phenotypes were downloaed from dbGaP on 12/14/2019.

### Files

- Read counts: `GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz`
- Private sample attributes: `phs000424.v8.pht002743.v8.p2.c1.GTEx_Sample_Attributes.GRU.txt`
- Private subject attributes: `phs000424.v8.pht002742.v8.p2.c1.GTEx_Subject_Phenotypes.GRU.txt`


### GTEx workflow

Github: [RNA-seq pipeline for the GTEx Consortium](https://github.com/broadinstitute/gtex-pipeline/tree/master/rnaseq)

- __Annotation__: GENCODE v26 (GRCh38/hg38)
- __SamToFastq__: BAM to FASTQ conversion
- __STAR__: spliced alignment of RNA sequence reads (v2.5.3a)
- __bamsync__: utility for transferring QC flags from the input BAM and for re-generating read group IDs
- __Picard__: MarkDuplicates: mark duplicate reads
- __RNA-SeQC__: QC metrics and gene-level expression quantification (v1.1.9)
- __RSEM__: transcript expression quantification (v1.3.0)

Therefore the count file from portal contains __RSEM__ gene-level estimated counts.

```{R setup} 
suppressPackageStartupMessages({
  library(yarn)
  library(stringr)
  library(Biobase)
  library(dplyr)
  library(rafalib)
  library(readr)
  library(ggplot2)
  library(ggpubr)
})
```

### Load
```{R}
v8.dir <- '~/cvd/v8'
count <- read.delim(gzfile(file.path(v8.dir, 'exprs/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz')), skip = 2, check.names = F)
sample <- read.delim(file.path(v8.dir, 'pheno/phs000424.v8.pht002743.v8.p2.c1.GTEx_Sample_Attributes.GRU.txt'), comment.char = '#')
subj <- read.delim(file.path(v8.dir, 'pheno/phs000424.v8.pht002742.v8.p2.c1.GTEx_Subject_Phenotypes.GRU.txt'), comment.char = '#')
```


### Creating ExpressionSet object

Merge subject and sample information:
```{R}
# extract gene & gene id
geneDF <- count[, 1:2]
length(unique(geneDF$Description)) # <56200
rownames(count) <- count[, 1]
count2 <- as.matrix(count[, -c(1,2)])

for (i in 1:nrow(problems(count))) {
	count2[problems(count)$row[i], problems(count)$col[i]] = 100000
}

# merge sample & subj
sample2 <- sample[sample$SAMPID %in% colnames(count2),]
sample2 <- cbind(SUBJID=str_extract(sample2$SAMPID, "^[^-]*-[^-]*"), sample2) # add subject id
subj2 <- subj[subj$SUBJID %in% sample2$SUBJID, ]

pdata <- merge(sample2, subj2, by="SUBJID", all=T)
rownames(pdata) <- pdata$SAMPID
```

#### Processing GENCODE file
```{bash, eval=F}
cd cvd/v8/anno/
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/gencode.v26.basic.annotation.gtf.gz
gunzip gencode.v26.basic.annotation.gtf.gz
cat gencode.v26.basic.annotation.gtf | grep -v ^## | awk -F'\t' '{if($3=="gene") print $9}' | sed -r 's/[^\"]*([\"][^\"]*[\"][,]?)[^\"]*/\1/g' | sed s'/\""/\t/g'|sed s'/\"//g'|cut -f1-3 > tmp.1.txt
cat gencode.v26.basic.annotation.gtf | grep -v ^## | awk -F'\t' -vOFS='\t' '{if($3 == "gene") print $1,$4,$5}' > tmp.2.txt
paste tmp.2.txt tmp.1.txt |awk -vOFS="\t" '{print $4,$1,$2,$3,$6,$5}' > gencode.v26.proc.txt
sed -i '1 i\ID\tchr\tstart\tend\tgeneNames\tbiotype' gencode.v26.proc.txt # add header
rm tmp.*
```

`gencode.v26.proc.txt` only includes columns: `ID`, `chr`, `start`, `end`, `geneName`, `biotype`.

Read the processed gencode file:
```{R}
anno <- read.delim(file=file.path(v8.dir, 'anno/gencode.v26.proc.txt'), header = T)
anno$chr <- gsub('^chr', '', anno$chr)
rownames(anno) <- anno$ID
anno <- anno[, -1]
```

### PC data

> We used a new linear model that doesn't need the PCs later on, so this block won't run.

In `GTEx_Analysis_v8_eQTL_covariates` directory, there are covariates for each tissue and need to collect all subjects' PC into one file.

#### Collect PCs
```{r, eval=F}
pc.dir <- '~/cvd/v8/geno/GTEx_Analysis_v8_eQTL_covariates'
all.cov <- list.files(pc.dir, full.names = T)
pc.subj <- data.frame()

for (f in all.cov) {
  tmp <- read.delim(f, header = T, check.names = F)
  ttmp <- as.data.frame(t(tmp[1:3,]))
  colnames(ttmp) <- unlist(ttmp[1,])
  ttmp <- ttmp[-1,]
  pc.subj <- rbind(pc.subj, ttmp[!rownames(ttmp) %in% rownames(pc.subj),])
}

nrow(pc.subj) # 838 subjects only
pc.subj$SUBJID <- rownames(pc.subj)
```


#### Sorting phenotype data
```{r}
# pdata2 <- merge(pdata, pc.subj, by="SUBJID", all=T)

# sort by sample ID
pdata <- pdata[order(pdata$SAMPID),]
rownames(pdata) <- pdata$SAMPID
count2 <- count2[, match(rownames(pdata), colnames(count2))]
all(rownames(pdata)==colnames(count2))
```


```{R, eval=T}
# create ExpressionSet
experimentData <- new("MIAME",
                      name = "Chi-Lam Poon",
                      lab="Cyclomics Lab",
                      contact="",
                      title="GTEx v8",
                      abstract="This is the object for GTEx studies, including expression and phenotype data.",
                      url="")
metadata <- data.frame(labelDescription = rep(NA, ncol(pdata)),
                       row.names = colnames(pdata))
phenoData <- new("AnnotatedDataFrame", data=pdata, varMetadata=metadata)
featureData <- new("AnnotatedDataFrame", data=anno[rownames(count2), ])

eset <- ExpressionSet(assayData = count2,
                      phenoData = phenoData,
                      experimentData = experimentData,
                      featureData = featureData)

dim(eset);length(unique(pData(eset)$SUBJID))
saveRDS(eset, file = file.path(v8.dir, 'exprs/gtex.v8.rds'))
```
